protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,545 @@
+            last_quiered: iters_len - 1,
+        }
+    }
+
+    pub fn on_failure(&mut self, peer: &PeerId) {
+        // All peer failures are reported to all queries and thus to all peer
+        // iterators. If this iterator never started a request to the given peer
+        // ignore the failure.
+        if let Some(index) = self.yielded_peers.get(peer) {
+            self.iters[*index].on_failure(peer)
+        }
+    }
+
+    pub fn on_success<I>(&mut self, peer: &PeerId, closer_peers: I)
+    where
+        I: IntoIterator<Item = PeerId>,
+    {
+        self.iters[self.yielded_peers[peer]].on_success(peer, closer_peers);
+    }
+
+    pub fn is_waiting(&self, peer: &PeerId) -> bool {
+        self.iters[self.yielded_peers[peer]].is_waiting(peer)
+    }
+
+    pub fn next(&mut self, now: Instant) -> PeersIterState {
+        let mut state = PeersIterState::Finished;
+
+        // Order in which to query the iterators to ensure fairness. Make sure
+        // to query the previously queried iterator last.
+        let iter_order = {
+            let mut all = (0..self.iters.len()).collect::<Vec<_>>();
+            let mut next_up = all.split_off(self.last_quiered + 1);
+            next_up.append(&mut all);
+            next_up
+        };
+
+        for i in &mut iter_order.into_iter() {
+            self.last_quiered = i;
+            let iter = &mut self.iters[i];
+
+            loop {
+                match iter.next(now) {
+                    PeersIterState::Waiting(None) => {
+                        match state {
+                            PeersIterState::Waiting(Some(_)) => {
+                                // [`ClosestDisjointPeersIter::next`] returns immediately once a
+                                // [`ClosestPeersIter`] yielded a peer. Thus this state is
+                                // unreachable.
+                                unreachable!();
+                            },
+                            PeersIterState::Waiting(None) => {}
+                            PeersIterState::WaitingAtCapacity => {
+                                state = PeersIterState::Waiting(None)
+                            }
+                            PeersIterState::Finished => state = PeersIterState::Waiting(None),
+                        };
+
+                        break;
+                    }
+                    PeersIterState::Waiting(Some(peer)) => {
+                        if self.yielded_peers.contains_key(&*peer) {
+                            // Another iterator already returned this peer. S/Kademlia requires each
+                            // peer to be only used on one path. Marking it as failed for this
+                            // iterator, asking it to return another peer in the next loop
+                            // iteration.
+                            let peer = peer.into_owned();
+                            iter.on_failure(&peer);
```

Comment:
Why do we have to fail this query, can't we just *not make the query in the first place*?

Judging by the documentation of `PeersIterState` it does sound like the overall framework doesn't give us this fine-grained control into the query process, is that right? It seems to be a very inelegant abstraction that is also overly restrictive - it only tells us when we are "waiting", it doesn't tell us when we actually receive a reply and let us adjust the reaction in response to the reply.

Reply:
> the peer iterators are pure state machines

This is not the only way of representing the logic as a pure state machine - I think this was where the previous discussion / explanation stalled. For some reason you & Max thought that I was confusing this approach with a non-pure approach, which I wasn't.

> They yield the candidate peers to query (via next()) and expect to get back the results via on_success and on_failure. 

OK, now this makes sense. Then this iterator works backwards from the typical iterator pattern, which is to yield outputs from `next` rather than inputs. A name like "RequestDriver" would be less ambiguous and clearer.

It's further confusing, because doing it this way around, `next` is actually redundant - you can simply merge its functionality into `on_success` and `on_failure`. You can make the decision on what the next peer-to-query is, as soon as you get the result back from a previous query, there is no need to separate out this functionality into `on_success` and then later `next`, which seems more like shoehorning the required functionality into a pre-chosen abstract design pattern (the "iterator").

This merged `on_success/next` is how I have implemented the multipaths algorithm [here](https://research.web3.foundation/en/latest/polkadot/networking/5-discovery.html) also as a pure state machine, and `recv_result` would correspond to the combination of `on_success` then `next`.

> Thus what is being done here is merely that the composite iterator that enforces "disjoint paths" via individual iterators [..]

This all makes immediate sense given the explanation of what the interface `next`, `on_success` and `on_failure` do. However, I'd suggest that this approach is subpar and leads to exactly the problem I mention [here](https://forum.web3.foundation/t/s-kademlia-and-multipaths/310), and that the multipaths approach (treating the whole query across all the peers as a single fat multipath, rather than d disjoint paths) avoids this bad case.

Reply:
> Then this iterator works backwards from the typical iterator pattern, which is to yield results from next rather than "new things to query". A name like "RequestDriver" would be less ambiguous and clearer.

I think this is just a matter of what preconceptions one has (apart from the fact that names are always contentious). When you say "[..] typical iterator pattern, which is to yield results [..]" you seem to have a certain idea of what a "typical iterator pattern" is and what "results" are and what not. To me, an iterator, in the general sense, is just anything that yields a sequence of values. The choice of name "Iterator" here does not stem from a desire to shoehorn the code into some abstract pattern, but only because the state of an (_iterative_) query is essentially a prioritised list of peers and an _iterative_ query _iterates_ over that list, thereby updating the list accordingly as results come in, eventually terminating.

> It's further confusing, because doing it this way around, next is actually redundant - you can simply merge its functionality into on_success and on_failure. You can make the decision on what the next peers-to-query are, as soon as you get the result back from a previous query, there is no need to separate out this functionality into on_success and then later next [..]

The decision on what the next peers-to-query are is essentially made when the callbacks are called, yes, but leaving the request for the next peer to another API call (here `next`) instead of returning it from these functions directly is mainly done because the separation of these steps integrates well with the overall callback-based approach to the libp2p `NetworkBehaviour`: When a response comes in on the network on some connection, a certain callback is invoked, which can in turn invoke `on_success` / `on_failure` on the corresponding query. Notably there is no option at this point to directly tell libp2p that another peer should be queried. Instead, libp2p regularly calls `NetworkBehaviour::poll` to ask the behaviour to "make progress" and as a result tell the underlying libp2p networking layer what to do next (e.g. query another peer). That is why `Kademlia::poll`, which implements `NetworkBehaviour::poll` calls `QueryPool::next()` which in turn calls `next()` on the underlying peer iterators. So, one reason for this split is simply the easy integration with the I/O layer in this way.

> [..] and that the multipaths approach (treating the whole query across all the peers as a single fat multipath, rather than d disjoint paths) avoids this bad case.

I still need to take a closer look at that.

Reply:
> there is no need to separate out this functionality into on_success and then later `next`, 

Although perhaps doing it this way allows you to work with TCP flow control a bit better, since you only call `next` when the stream unblocks, and still want to receive results in the meantime. It's not clear if that was the purpose behind the original design though.

Reply:
> When you say "[..] typical iterator pattern, which is to yield results [..]" you seem to have a certain idea of what a "typical iterator pattern" is and what "results" are and what not. To me, an iterator, in the general sense, is just anything that yields a sequence of values.

Yes I can see this perspective too. However when trying to quickly understand new code you often have to take shortcuts and "iterator" suggests a producer that *doesn't care what the consumer does* with the yielded value - whereas here the behaviour of the consumer also very much affects the producer, via `on_success` and `on_failure`. This would be more completely-described by the term "coroutine" - although you are right that an iterator technically doesn't prohibit this, it does connotate it to a human reading the code.

> The choice of name "Iterator" here does not stem from a desire to shoehorn the code into some abstract pattern, but only because the state of an (iterative) query is essentially a prioritised list of peers and an iterative query iterates over that list, [..]

Yeah fair enough, I can see the resemblance from this p2p usage of the terminology, esp in terms of "iterative" vs "recursive" queries. The term "iteration" is overloaded a bit.